rllab/envs/gym_util/space_util.py

```diff
@@ -0,0 +1,151 @@
+    'new_tensor_variable'
+]
+
+
+def bounds(space):
+    if isinstance(space, gym.spaces.Box):
+        return space.low, space.high
+    else:
+        raise NotImplementedError
+
+
+def components(space):
+    if isinstance(space, gym.spaces.Tuple):
+        return self.spaces
+    else:
+        raise NotImplementedError
+
+
+def default_value(space):
+    if isinstance(space, gym.spaces.Discrete):
+        return 0
+    else:
+        return NotImplementedError
+
+
+def flat_dim(space):
+    if isinstance(space, gym.spaces.Box):
+        return np.prod(space.low.shape)
+    elif isinstance(space, gym.spaces.Discrete):
+        return space.n
+    elif isinstance(space, gym.spaces.Tuple):
+        return np.sum([flat_dim(x) for x in space.spaces])
+    else:
+        raise NotImplementedError
+
+
+def flatten(space, obs):
+    if isinstance(space, gym.spaces.Box):
+        return np.asarray(obs).flatten()
+    elif isinstance(space, gym.spaces.Discrete):
+        return special.to_onehot(obs, space.n)
+    elif isinstance(space, gym.spaces.Tuple):
+        return np.concatenate(
+            [flatten(xi, c) for c, xi in zip(space.spaces, obs)])
+    else:
+        raise NotImplementedError
+
+
+def flatten_n(space, obs):
+    if isinstance(space, gym.spaces.Box):
+        obs = np.asarray(obs)
+        return obs.reshape((obs.shape[0], -1))
+    elif isinstance(space, gym.spaces.Discrete):
+        return special.to_onehot_n(obs, space.n)
+    elif isinstance(space, gym.spaces.Tuple):
+        obs_regrouped = [[obs[i] for o in obs] for i in range(len(obs[0]))]
+        flat_regrouped = [
+            flatten_n(c, oi) for c, oi in zip(space.spaces, obs_regrouped)
+        ]
+        return np.concatenate(flat_regrouped, axis=-1)
+    else:
+        raise NotImplementedError
+
+
+def log_diagnostics(space, paths):
+    pass
+
+
+def sample(space):
+    if isinstance(space, gym.spaces.Tuple):
+        return tuple(x.sample() for x in self.spaces)
+    else:
+        raise NotImplementedError
+
+
+def unflatten(space, obs):
+    if isinstance(space, gym.spaces.Box):
+        return np.asarray(obs).reshape(space.shape)
+    elif isinstance(space, gym.spaces.Discrete):
+        return special.from_onehot(obs)
+    elif isinstance(space, gym.spaces.Tuple):
+        dims = [flat_dim(c) for c in space.spaces]
+        flat_xs = np.split(x, np.cumsum(dims)[:-1])
+        return tuple(unflatten(c, xi) for c, xi in zip(self.spaces, flat_xs))
+    else:
+        raise NotImplementedError
+
+
+def unflatten_n(space, obs):
+    if isinstance(space, gym.spaces.Box):
+        obs = np.asarray(obs)
+        return obs.reshape((obs.shape[0], ) + space.shape)
+    elif isinstance(space, gym.spaces.Discrete):
+        return special.from_onehot_n(obs)
+    elif isinstance(space, gym.spaces.Tuple):
+        dims = [flat_dim(c) for c in self.spaces]
+        flat_xs = np.split(xs, np.cumsum(dims)[:-1], axis=-1)
```

Comment:
Same as unflatten(space, obs).. also line 107, there is not self here because is not an object.